### PR TITLE
Update handling of AccessPolicy write criteria and add docs

### PR DESCRIPTION
**Pull request by mattwiller**

### CODE REVIEW

File: packages/docs/static/data/medplumDefinitions/accesspolicy.json [hunk 1]
```diff
@@ -239,48 +239,16 @@
...
         }
       ],
-      "path": "AccessPolicy.resource.writeCriteria.post",
+      "path": "AccessPolicy.resource.writeCriteria",
       "min": 0,
-      "max": "1",
+      "max": "*",
       "short": "",
-      "definition": "Invariant to check against the state of the resource with modifications tentatively applied.",
+      "definition": "Invariant that must be satisfied for the resource to be written.  Can include %before and %after placeholders to refer to the resource before and after the updates are applied.",
```
Comment (rahul1) [inline packages/docs/static/data/medplumDefinitions/accesspolicy.json:251]: What was your thinking behind making this an array field? Couldn't we get the same functionality by a single long expression with ORs and ANDs

Comment (mattwiller) [reply packages/docs/static/data/medplumDefinitions/accesspolicy.json:251]: We could, but when I tried that I thought even with two conditions it got unwieldy.  Having the field be an array allows for more granular management of the individual conditions, and frees users from needing to worry about parentheses and operator precedence order most of the time

Comment (rahul1) [reply packages/docs/static/data/medplumDefinitions/accesspolicy.json:251]: I see. It seems like you took inspiration from the profile invariants? 

My concern is about implicit behavior. Once you have an array of conditions, you need to know if they're OR'd or AND'd together, so either have to reference the docs, or have another field specify the behavior. That's what `enableWhen` does in `Questionnaire`. But maybe if they're all _invariants_, it's clear from context that they must all be satisfied?

Comment (mattwiller) [reply packages/docs/static/data/medplumDefinitions/accesspolicy.json:251]: I think that it should ideally be clear form context that all the criteria must be satisfied: would it help if we renamed the field to e.g. `writeConstraint` or `writeInvariant`?

File: packages/docs/static/data/medplumDefinitions/accesspolicy.json [hunk 2]
```diff
@@ -239,48 +239,16 @@
...
-      "definition": "Invariant to check against the state of the resource before modification.",
-      "comment": "",
-      "inherited": false
-    },
-    {
-      "name": "post",
-      "depth": 3,
       "types": [
         {
           "datatype": "string"
```
[the remark this answers is not in the document]
Comment (mattwiller) [reply packages/docs/static/data/medplumDefinitions/accesspolicy.json:244]: I like the idea of using `Expression` to disambiguate between `text/fhirpath` and `application/x-fhir-query` inputs; I can go ahead and make that change

File: packages/docs/static/data/medplumDefinitions/accesspolicy.json (hunk 1 above)
Comment (rahul1) [reply packages/docs/static/data/medplumDefinitions/accesspolicy.json:252]: Yeah I think that's a good suggestion. `writeConstraint` maps nicely to the  overall `constraint` defined for [ElementDefinition](https://build.fhir.org/elementdefinition-definitions.html#ElementDefinition.constraint)

File: packages/definitions/dist/fhir/r4/fhir.schema.json [hunk 3]
```diff
@@ -61917,8 +61917,11 @@
           "type": "array"
         },
         "writeCriteria": {
```
Comment (rahul1) [inline packages/definitions/dist/fhir/r4/fhir.schema.json:61919]: @mattwiller I believe this needs to be updated as well

File: packages/docs/docs/auth/access-control.md [hunk 4]
```diff
@@ -119,6 +119,40 @@ The following policy uses a FHIR Search Query to grant access only to Coverage w
 
 Note that in this implementation access policy, only the `Coverage` resources that have `payor=Organization/123` will be visible.
 
+Criteria for writes to a resource can also be specified using [FHIRPath expressions][fhirpath] in the `AccessPolicy.resource.writeCriteria` field. These expressions may contain the special variable `%before` to refer to the resource as it existed before the write. Any property accesses will by default refer to the resource as it exists with updates applied, but the `%after` variable is also provided for convenience.
```
Comment (rahul1) [inline packages/docs/docs/auth/access-control.md:122]: AccessPolicy.resource.writeCriteria -> AccessPolicy.resource.writeConstraint

File: packages/docs/docs/auth/access-control.md [hunk 5]
```diff
@@ -119,6 +119,40 @@ The following policy uses a FHIR Search Query to grant access only to Coverage w
...
+On resource creation, `%before` will be undefined, so any expressions that refer to `%before` must account for this case. To select only updates or only creates, prefix the criteria with `%before.exists() implies` or `%before.exists().not() implies` respectively.
+
+:::
+
+For example, an access policy with write criteria could be used to manage state transitions by prohibiting changing the status once a resource is marked as `final`, and ensure that a `subject` is set when the resource is finalized:
+
+```json
+{
+  "resourceType": "AccessPolicy",
+  "name": "Write Criteria Access Policy",
```
Comment (rahul1) [inline packages/docs/docs/auth/access-control.md:135]: "Write Constraint Access Policy"

File: packages/docs/docs/auth/access-control.md (hunk 4 above)
Comment (rahul1) [inline packages/docs/docs/auth/access-control.md:122]: "Constraints on"

File: packages/docs/docs/auth/access-control.md [hunk 6]
```diff
@@ -119,6 +119,40 @@ The following policy uses a FHIR Search Query to grant access only to Coverage w
 
 Note that in this implementation access policy, only the `Coverage` resources that have `payor=Organization/123` will be visible.
 
+Criteria for writes to a resource can also be specified using [FHIRPath expressions][fhirpath] in the `AccessPolicy.resource.writeCriteria` field. These expressions may contain the special variable `%before` to refer to the resource as it existed before the write. Any property accesses will by default refer to the resource as it exists with updates applied, but the `%after` variable is also provided for convenience.
+
+:::tip
+
+On resource creation, `%before` will be undefined, so any expressions that refer to `%before` must account for this case. To select only updates or only creates, prefix the criteria with `%before.exists() implies` or `%before.exists().not() implies` respectively.
```
Comment (rahul1) [inline packages/docs/docs/auth/access-control.md:126]: "criteria" -> "constraint"

File: packages/docs/docs/auth/access-control.md [hunk 7]
```diff
@@ -119,6 +119,40 @@ The following policy uses a FHIR Search Query to grant access only to Coverage w
...
+
+:::
+
+For example, an access policy with write criteria could be used to manage state transitions by prohibiting changing the status once a resource is marked as `final`, and ensure that a `subject` is set when the resource is finalized:
```
Comment (rahul1) [inline packages/docs/docs/auth/access-control.md:130]: "criteria" -> "constraint"

File: packages/docs/static/data/medplumDefinitions/accesspolicy.json [hunk 8]
```diff
@@ -241,46 +241,15 @@
       "depth": 2,
       "types": [
         {
-          "datatype": "BackboneElement"
+          "datatype": "Expression",
+          "documentLocation": "datatype"
         }
       ],
       "path": "AccessPolicy.resource.writeCriteria",
```
Comment (rahul1) [inline packages/docs/static/data/medplumDefinitions/accesspolicy.json:248]: "writeCriteria" -> "writeConstraint"

File: packages/fhirtypes/dist/AccessPolicy.d.ts [hunk 9]
```diff
@@ -131,25 +132,9 @@ export interface AccessPolicyResource {
...
-   * modification.
-   */
-  pre?: string;
-
-  /**
-   * Invariant to check against the state of the resource with
-   * modifications tentatively applied.
-   */
-  post?: string;
+  writeCriteria?: Expression[];
```
Comment (rahul1) [inline packages/fhirtypes/dist/AccessPolicy.d.ts:139]: need to run the type generator again: 
"writeCriteria" -> "writeConstraint"

File: packages/server/src/fhir/accesspolicy.test.ts [hunk 10]
```diff
@@ -1291,10 +1291,16 @@ describe('AccessPolicy', () => {
       resource: [
         {
           resourceType: 'Observation',
-          writeCriteria: {
-            pre: `status != 'final'`,
-            post: `status = 'final' implies subject.exists()`,
-          },
+          writeCriteria: [
```
Comment (rahul1) [inline packages/server/src/fhir/accesspolicy.test.ts:1294]: "writeCriteria" -> "writeConstraint"

File: packages/server/src/fhir/repo.ts [hunk 11]
```diff
@@ -1550,33 +1550,31 @@ export class Repository extends BaseRepository implements FhirRepository {
...
+  private isResourceWriteable(previous: Resource | undefined, current: Resource): boolean {
+    const matchingPolicy = satisfiedAccessPolicy(current, AccessPolicyInteraction.UPDATE, this.context.accessPolicy);
     if (!matchingPolicy) {
       return false;
-    } else if (matchingPolicy?.writeCriteria?.post) {
-      const invariant = evalFhirPathTyped(matchingPolicy.writeCriteria.post, [
-        { type: resource.resourceType, value: resource },
-      ]);
-      return invariant.length === 1 && invariant[0].value === true;
+    } else if (matchingPolicy?.writeCriteria) {
```
Comment (rahul1) [inline packages/server/src/fhir/repo.ts:1566]: This code will need to be updated: 
"writeCriteria" -> "writeConstraint"

**Review by rahul1 — CHANGES_REQUESTED** [review]
Substance looks good, but a bunch of places where "writeCriteria" needs to be replaced with "writeConstraint"

File: packages/definitions/dist/fhir/r4/profiles-medplum.json [hunk 12]
```diff
@@ -2643,49 +2643,27 @@
...
               "min" : 0,
-              "max" : "1"
-            }
+              "max" : "*"
+            },
+            "constraint": [
+              {
+                "key": "axp-1",
+                "severity": "error",
+                "human": "Write criteria expressions must be literal text/fhirpath strings",
```
Comment (rahul1) [inline packages/definitions/dist/fhir/r4/profiles-medplum.json:2663]: "Write criteria expressions " -> "Write constraint expressions..."

File: packages/server/src/fhir/repo.ts [hunk 13]
```diff
@@ -1550,33 +1550,31 @@ export class Repository extends BaseRepository implements FhirRepository {
...
       return false;
-    } else if (matchingPolicy?.writeCriteria?.post) {
-      const invariant = evalFhirPathTyped(matchingPolicy.writeCriteria.post, [
-        { type: resource.resourceType, value: resource },
-      ]);
-      return invariant.length === 1 && invariant[0].value === true;
+    } else if (matchingPolicy?.writeConstraint) {
+      return matchingPolicy.writeConstraint.every((criteria) => {
+        const invariant = evalFhirPathTyped(
+          criteria.expression as string,
```
Comment (rahul1) [inline packages/server/src/fhir/repo.ts:1569]: `criteria` -> `contraint`

**Review by rahul1 — CHANGES_REQUESTED** [review]

**Review by rahul1 — APPROVED** [review]

**Pull request merged**